src/hooks/useTypedSignTrustCredential.ts

```diff
@@ -0,0 +1,118 @@
+
+import { generateTCTypedSignPayload } from '../utils';
+
+export enum SignatureErrorTypes {
+  Invalid = 'invalid',
+  Error = 'error',
+}
+
+export type SignatureError = {
+  type: SignatureErrorTypes;
+  message: string;
+};
+
+/**
+ * A Hook to sign TrustCredential typed data.
+ *
+ * @returns The signature and the function to sign the message. isLoading and isVerified are used to indicate the status of the signature.
+ */
+export function useTypedSignTrustCredential() {
+  const client = usePublicClient();
+  const chainId = useChainId();
+  const { address } = useAccount();
+  const { data: signature, signTypedData } = useSignTypedData();
+
+  const [trustCredentialTypedData, setTrustCredentialTypedData] =
+    useState<SignTypedDataArgs>();
+  // above useSignTypedData have isLoading, but for some reason, it is not working well with loading state, use following instead
+  const [isLoading, setIsLoading] = useState(false);
+  const [isVerified, setIsVerified] = useState(false);
+  // Type [key_4: string] is defined inside the  SignTypedDataArgs in Wagmi
+  const [payload, setPayload] = useState<{ [key_4: string]: unknown }>();
+  const [signatureError, setSignatureError] = useState<SignatureError>();
+
+  const submitTypedSignRequest = useCallback(
+    (subjectAddress: string, isAdd: boolean) => {
+      if (!address) {
+        return;
+      }
+      setIsLoading(true);
+      const tcTypedSignData = generateTCTypedSignPayload(
+        address,
+        subjectAddress,
+        chainId,
+        isAdd,
+      );
+      signTypedData(tcTypedSignData);
+      setTrustCredentialTypedData(tcTypedSignData);
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [address],
+  );
+
+  useEffect(() => {
+    if (
+      address &&
+      trustCredentialTypedData !== undefined &&
+      signature !== undefined
+    ) {
+      const { domain, types, primaryType, message } = trustCredentialTypedData;
+
+      client
+        .verifyTypedData({
+          address,
+          domain,
+          types,
+          primaryType,
+          message,
+          signature,
+        })
+        .then((res) => {
+          if (res) {
+            setIsLoading(false);
+
+            setIsVerified(res);
+
+            const tcPayload = trustCredentialTypedData.message;
+            tcPayload.proof = {
+              type: 'EthereumEip712Signature2021',
+              proofValue: signature,
+            };
+            setPayload(tcPayload);
+          } else {
+            setSignatureError({
+              type: SignatureErrorTypes.Invalid,
+              message: 'Invalid signature',
+            });
+            setIsVerified(false);
+            setIsLoading(false);
+          }
+        })
+        .catch((error) => {
+          setSignatureError({
+            type: SignatureErrorTypes.Error,
+            message: error.message,
+          });
+          setIsVerified(false);
+          setIsLoading(false);
+        });
```

Comment:
```suggestion
      setIsLoading(true)
      client
        .verifyTypedData({
          address,
          domain,
          types,
          primaryType,
          message,
          signature,
        })
        .then((res) => {
          if (res) {
            setIsVerified(res);
            const tcPayload = trustCredentialTypedData.message;
            tcPayload.proof = {
              type: 'EthereumEip712Signature2021',
              proofValue: signature,
            };
            setPayload(tcPayload);
          } else {
            setSignatureError({
              type: SignatureErrorTypes.Invalid,
              message: 'Invalid signature',
            });
            setIsVerified(false);
          }
        })
        .catch((error) => {
          setSignatureError({
            type: SignatureErrorTypes.Error,
            message: error.message,
          });
          setIsVerified(false);
        }).finally(() => {
            setIsLoading(false);
        });
```

Reply:
Sorry, the `setIsLoading(false)` here to each promise then, catch block here to indicate the process have completed finish, so we can't put the `setIsLoading` on the top.  

the isLoading will be set to `true` when `submitTypedSignRequest` run to indicate `signing` process start, this will control the `sign` button disabled. and after user sign, `useEffect` will run to verify signature, and then set the payload or set the error if error occur, and set isLoading to false to enable the button again.